File: LegionGalaxy/LegionGalaxy/Models/GameModels.swift
import Foundation

enum BuildingType: String, CaseIterable, Identifiable {
    case colonyCenter = "Colony Center"
    case resourceCollector = "Resource Collector"
    case barracks = "Barracks"
    case farm = "Farm"
    case house = "House"
    case researchCenter = "Research Center"
    case excellorantFactory = "Excellorant Factory"
    case spaceport = "Spaceport"

    var id: String { rawValue }

    var icon: String {
        switch self {
        case .colonyCenter: return "building.columns.fill"
        case .resourceCollector: return "cube.fill"
        case .barracks: return "shield.lefthalf.filled"
        case .farm: return "leaf.fill"
        case .house: return "house.fill"
        case .researchCenter: return "atom"
        case .excellorantFactory: return "bolt.fill"
        case .spaceport: return "airplane.departure"
        }
    }

    var summary: String {
        switch self {
        case .colonyCenter: return "Core hub of your base. Gates the max level of every other building."
        case .resourceCollector: return "Gathers construction materials. Upgrade to increase yield."
        case .barracks: return "Trains Infantry, Tanks, and Mounted Units from your population."
        case .farm: return "Produces food to sustain and grow your population."
        case .house: return "Increases population capacity for workers and troops."
        case .researchCenter: return "Unlocks and improves technologies, including ship upgrades."
        case .excellorantFactory: return "Produces Excellorant, a boost resource that speeds up ship factories, barracks, resource collectors, and farms."
        case .spaceport: return "Build starships and launch them into your solar system. Unlocked after conquering 3 of 5 continents."
        }
    }
}

struct Building: Identifiable {
    let id = UUID()
    let type: BuildingType
    var level: Int

    var upgradeCost: Int { 100 * level * level }
}

enum AlienThreatLevel: String {
    case dormant = "Dormant"
    case guarded = "Guarded"
    case fortified = "Fortified"
}

struct Continent: Identifiable {
    let id = UUID()
    let name: String
    var conquered: Bool
    var threat: AlienThreatLevel
    var buildings: [Building]
}

struct Planet: Identifiable {
    let id = UUID()
    let name: String
    let isAgriWorld: Bool
    let isHome: Bool
    var continents: [Continent]

    var conqueredCount: Int { continents.filter(\.conquered).count }
    var spaceportUnlocked: Bool { conqueredCount >= 3 }
}

enum TroopType: String, CaseIterable {
    case infantry = "Infantry"
    case tanks = "Tanks"
    case mounted = "Mounted Units"

    var icon: String {
        switch self {
        case .infantry: return "person.fill"
        case .tanks: return "shield.fill"
        case .mounted: return "hare.fill"
        }
    }

    var role: String {
        switch self {
        case .infantry: return "All-purpose ground unit"
        case .tanks: return "Defensive specialist (late game)"
        case .mounted: return "Fast-attack raider"
        }
    }
}

enum ShipType: String, CaseIterable {
    case fighter = "Fighter"
    case hauler = "Hauler"
    case cruiser = "Cruiser"
    case probe = "Probe"
    case dreadnaught = "Dreadnaught"

    var role: String {
        switch self {
        case .fighter: return "Cheap and used in large numbers"
        case .hauler: return "Carries cargo and ground units"
        case .cruiser: return "Slower, general-purpose space combat ship"
        case .probe: return "Recon other planets and their surface activity"
        case .dreadnaught: return "Very expensive capital ship for fleet battles"
        }
    }
}

/// Sample game-state factory used to seed the UI with representative data.
enum GameWorld {
    static func sampleHomePlanet() -> Planet {
        let starterBuildings: [Building] = [
            Building(type: .colonyCenter, level: 1),
            Building(type: .resourceCollector, level: 1),
            Building(type: .barracks, level: 1),
            Building(type: .farm, level: 1),
            Building(type: .house, level: 1),
            Building(type: .researchCenter, level: 1),
            Building(type: .excellorantFactory, level: 1)
        ]
        let continents = [
            Continent(name: "Continent I — Aurelia", conquered: true, threat: .dormant, buildings: starterBuildings),
            Continent(name: "Continent II — Voss", conquered: false, threat: .guarded, buildings: []),
            Continent(name: "Continent III — Kethra", conquered: false, threat: .guarded, buildings: []),
            Continent(name: "Continent IV — Drannix", conquered: false, threat: .fortified, buildings: []),
            Continent(name: "Continent V — Solhaven", conquered: false, threat: .fortified, buildings: [])
        ]
        return Planet(name: "New Terra", isAgriWorld: false, isHome: true, continents: continents)
    }

    static func sampleSolarSystem() -> [Planet] {
        [
            sampleHomePlanet(),
            Planet(name: "Verdant Expanse", isAgriWorld: true, isHome: false, continents: []),
            Planet(name: "Cinder Hollow", isAgriWorld: false, isHome: false, continents: []),
            Planet(name: "Glacian Reach", isAgriWorld: false, isHome: false, continents: []),
            Planet(name: "Obsidian Drift", isAgriWorld: false, isHome: false, continents: [])
        ]
    }
}

File: Unity/LegionGalaxy/Assets/Scripts/Models/GameModels.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace LegionGalaxy.Models
{
    public enum BuildingType
    {
        ColonyCenter, ResourceCollector, Barracks, Farm, House,
        ResearchCenter, ExcellorantFactory, Spaceport
    }

    public static class BuildingInfo
    {
        public static string DisplayName(BuildingType t) => t switch
        {
            BuildingType.ColonyCenter => "Colony Center",
            BuildingType.ResourceCollector => "Resource Collector",
            BuildingType.Barracks => "Barracks",
            BuildingType.Farm => "Farm",
            BuildingType.House => "House",
            BuildingType.ResearchCenter => "Research Center",
            BuildingType.ExcellorantFactory => "Excellorant Factory",
            BuildingType.Spaceport => "Spaceport",
            _ => t.ToString()
        };

        public static string Summary(BuildingType t) => t switch
        {
            BuildingType.ColonyCenter => "Core hub of your base. Gates the max level of every other building.",
            BuildingType.ResourceCollector => "Gathers construction materials. Upgrade to increase yield.",
            BuildingType.Barracks => "Trains Infantry, Tanks, and Mounted Units from your population.",
            BuildingType.Farm => "Produces food to sustain and grow your population.",
            BuildingType.House => "Increases population capacity for workers and troops.",
            BuildingType.ResearchCenter => "Unlocks and improves technologies, including ship upgrades.",
            BuildingType.ExcellorantFactory => "Produces Excellorant, a boost resource that speeds up ship factories, barracks, resource collectors, and farms.",
            BuildingType.Spaceport => "Build starships and launch them into your solar system. Unlocked after conquering 3 of 5 continents.",
            _ => ""
        };
    }

    [Serializable]
    public class Building
    {
        public BuildingType Type;
        public int Level = 1;
        public int UpgradeCost => 100 * Level * Level;

        public Building(BuildingType type, int level = 1) { Type = type; Level = level; }
    }

    public enum AlienThreatLevel { Dormant, Guarded, Fortified }

    [Serializable]
    public class Continent
    {
        public string Name;
        public bool Conquered;
        public AlienThreatLevel Threat;
        public List<Building> Buildings = new();

        public Continent(string name, bool conquered, AlienThreatLevel threat, List<Building> buildings = null)
        {
            Name = name; Conquered = conquered; Threat = threat;
            Buildings = buildings ?? new List<Building>();
        }
    }

    [Serializable]
    public class Planet
    {
        public string Name;
        public bool IsAgriWorld;
        public bool IsHome;
        public List<Continent> Continents = new();

        public int ConqueredCount => Continents.Count(c => c.Conquered);
        public bool SpaceportUnlocked => ConqueredCount >= 3;

        public Planet(string name, bool isAgriWorld, bool isHome, List<Continent> continents = null)
        {
            Name = name; IsAgriWorld = isAgriWorld; IsHome = isHome;
            Continents = continents ?? new List<Continent>();
        }
    }

    public enum TroopType { Infantry, Tanks, MountedUnits }

    public static class TroopInfo
    {
        public static string DisplayName(TroopType t) => t switch
        {
            TroopType.Infantry => "Infantry",
            TroopType.Tanks => "Tanks",
            TroopType.MountedUnits => "Mounted Units",
            _ => t.ToString()
        };

        public static string Role(TroopType t) => t switch
        {
            TroopType.Infantry => "All-purpose ground unit",
            TroopType.Tanks => "Defensive specialist (late game)",
            TroopType.MountedUnits => "Fast-attack raider",
            _ => ""
        };
    }

    public enum ShipType { Fighter, Hauler, Cruiser, Probe, Dreadnaught }

    public static class ShipInfo
    {
        public static string Role(ShipType t) => t switch
        {
            ShipType.Fighter => "Cheap and used in large numbers",
            ShipType.Hauler => "Carries cargo and ground units",
            ShipType.Cruiser => "Slower, general-purpose space combat ship",
            ShipType.Probe => "Recon other planets and their surface activity",
            ShipType.Dreadnaught => "Very expensive capital ship for fleet battles",
            _ => ""
        };
    }

    /// Seeds representative game state for the UI scaffold.
    public static class GameWorld
    {
        public static Planet SampleHomePlanet()
        {
            var starterBuildings = new List<Building>
            {
                new(BuildingType.ColonyCenter),
                new(BuildingType.ResourceCollector),
                new(BuildingType.Barracks),
                new(BuildingType.Farm),
                new(BuildingType.House),
                new(BuildingType.ResearchCenter),
                new(BuildingType.ExcellorantFactory)
            };

            var continents = new List<Continent>
            {
                new("Continent I — Aurelia", true, AlienThreatLevel.Dormant, starterBuildings),
                new("Continent II — Voss", false, AlienThreatLevel.Guarded),
                new("Continent III — Kethra", false, AlienThreatLevel.Guarded),
                new("Continent IV — Drannix", false, AlienThreatLevel.Fortified),
                new("Continent V — Solhaven", false, AlienThreatLevel.Fortified)
            };

            return new Planet("New Terra", false, true, continents);
        }

        public static List<Planet> SampleSolarSystem() => new()
        {
            SampleHomePlanet(),
            new Planet("Verdant Expanse", true, false),
            new Planet("Cinder Hollow", false, false),
            new Planet("Glacian Reach", false, false),
            new Planet("Obsidian Drift", false, false)
        };
    }
}
